ledger/test/test_compact_serializer.py

```diff
@@ -0,0 +1,228 @@
+
+    assert {"f1": "v1", "f2": "v2", "f3": 3, "f4": 4.0, "f5": True} == \
+            serializer.deserialize(b"v1|v2|3|4.0|True")
+
+
+def testSerializeToFields():
+    serializer = CompactSerializer(fields)
+
+    newFields = OrderedDict([
+        ("ff1", (str, str)),
+        ("ff2", (str, str)),
+        ("ff3", (str, int))
+    ])
+    assert b"v1|v2|3" == \
+           serializer.serialize(
+               {"ff1": "v1", "ff2": "v2", "ff3": 3},
+               fields=newFields
+           )
+
+def testDeserializeFromString():
+    serializer = CompactSerializer(fields)
+
+    assert {"f1": "v1", "f2": "v2", "f3": 3, "f4": 4.0, "f5": True} == \
+            serializer.deserialize("v1|v2|3|4.0|True")
+
+def testDeserializForFields():
+    serializer = CompactSerializer(fields)
+
+    newFields = OrderedDict([
+        ("ff1", (str, str)),
+        ("ff2", (str, str)),
+        ("ff3", (str, int))
+    ])
+    assert {"ff1": "v1", "ff2": "v2", "ff3": 3} == \
+            serializer.deserialize("v1|v2|3", fields=newFields)
+
+
+def testSerializeLessFields():
+    serializer = CompactSerializer(fields)
+
+    assert b"|v1|2||" == serializer.serialize({"f2": "v1", "f3": 2})
+    assert b"v1||||" == serializer.serialize({"f1": "v1"})
+    assert b"||||" == serializer.serialize({})
+    assert b"||||3" == serializer.serialize({"f5": 3})
+    assert b"v1||||3" == serializer.serialize({"f1": "v1", "f5": 3})
+
+
+def testDeserializeLessFields():
+    serializer = CompactSerializer(fields)
+
+    assert {"f1": None, "f2": "v1", "f3": 2, "f4": None, "f5": None} == serializer.deserialize(b"|v1|2||")
+    assert {"f1": "v1", "f2": None, "f3": None, "f4": None, "f5": None} == serializer.deserialize(b"v1||||")
+    assert {"f1": None, "f2": None, "f3": None, "f4": None, "f5": None} == serializer.deserialize(b"|||||")
+    assert {"f1": None, "f2": None, "f3": None, "f4": None, "f5": True} == serializer.deserialize(b"||||True" )
+    assert {"f1": "v1", "f2": None, "f3": None, "f4": None, "f5": False} == serializer.deserialize(b"v1||||False")
+
+
+def testSerializeLessFieldsWithNone():
+    serializer = CompactSerializer(fields)
+
+    assert b"|v1|2||" == serializer.serialize({"f1": None, "f2": "v1", "f3": 2})
+    assert b"v1||||" == serializer.serialize({"f1": "v1", "f2": None, "f3": None, "f4": None, "f5": None})
+    assert b"||||" == serializer.serialize({"f1": None, "f2": None, "f3": None, "f4": None, "f5": None})
+    assert b"||||3" == serializer.serialize({"f1": None, "f5": 3})
+    assert b"||||3" == serializer.serialize({"f1": None, "f2": None, "f3": None, "f4": None, "f5": 3})
+    assert b"v1||||3" == serializer.serialize({"f1": "v1", "f2": None, "f3": None, "f4": None, "f5": 3})
+    assert b"v1||||3" == serializer.serialize({"f1": "v1", "f3": None, "f5": 3})
+
+def testSerializeInAnyOrder():
+    serializer = CompactSerializer(fields)
+
+    assert b"|v1|2||" == serializer.serialize({"f3": 2, "f2": "v1"})
```

Comment:
Using OrderedDict for input now.
The test checks that the serialized data is the same regardless of the input data order.